### PR TITLE
[ADD] l10n_no_edi: added support for EHF3

### PR DESCRIPTION
**Pull request by bfr-o**
EHF3 is the Norvegian standard for Peppol. This module generates xml (based on BIS3 with NO country profile) to be submitted on the Peppol network.

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/l10n_no/models/res_partner.py [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# coding: utf-8
+from odoo import fields, models
+
+
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+
+    l10n_no_bronnoysund_number = fields.Char(string='Bronnoysund company registry', size=9)  # Enhetsregisteret ved Bronnoysundregisterne
```
Comment (smetl) [inline addons/l10n_no/models/res_partner.py:8]: the comment is useless except if turned in english

File: addons/l10n_no_edi/data/ehf_3_template.xml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+                  primary="True">
+            <xpath expr="//*[local-name()='PartyTaxScheme']" position="after">
+                <cac:PartyTaxScheme xmlns="urn:oasis:names:specification:ubl:schema:xsd:Invoice-2"
+                xmlns:cac="urn:oasis:names:specification:ubl:schema:xsd:CommonAggregateComponents-2"
+                xmlns:cbc="urn:oasis:names:specification:ubl:schema:xsd:CommonBasicComponents-2">
+                    <cbc:CompanyID>Foretaksregisteret</cbc:CompanyID>
+                    <cac:TaxScheme>
+                        <cbc:ID>TAX</cbc:ID>
+                    </cac:TaxScheme>
+                </cac:PartyTaxScheme>
```
Comment (smetl) [inline addons/l10n_no_edi/data/ehf_3_template.xml:15]: This could be added in generic when the withholding tax task will be merged since there will be a `family='vat'` field on taxes I think

File: addons/account_edi_ubl/data/ubl_templates.xml [hunk 3]
```diff
@@ -117,7 +117,7 @@
                 <cbc:ID t-esc="record.name"/>
                 <cbc:IssueDate t-esc="record.invoice_date"/>
                 <cbc:InvoiceTypeCode t-esc="type_code"/>
-                <cbc:Note t-esc="record.narration"/>
+                <cbc:Note t-esc="note"/>
```
Comment (smetl) [inline addons/account_edi_ubl/data/ubl_templates.xml:120]: missing `t-if`

File: addons/l10n_no_edi/data/ehf_3_template.xml (hunk 2 above)
Comment (bfr-o) [reply addons/l10n_no_edi/data/ehf_3_template.xml:15]: Yep !

File: addons/account_edi_ubl/models/account_edi_format.py [hunk 4]
```diff
@@ -195,6 +195,7 @@ def format_monetary(amount):
             'type_code': 380 if invoice.move_type == 'out_invoice' else 381,
             'payment_means_code': 42 if invoice.journal_id.bank_account_id else 31,
             'bank_account': invoice.partner_bank_id,
+            'note': html2plaintext(invoice.narration) if invoice.narration else False,
```
Comment (smetl) [inline addons/account_edi_ubl/models/account_edi_format.py:198]: this is already fixed in master, right?

File: addons/account_edi_ubl/models/account_edi_format.py [hunk 5]
```diff
@@ -257,10 +258,3 @@ def _post_invoice_edi(self, invoices):
             attachment = self._export_ubl(invoice)
             res[invoice] = {'attachment': attachment}
         return res
-
-    def _is_embedding_to_invoice_pdf_needed(self):
-        # OVERRIDE
-        self.ensure_one()
-        if self.code != 'ubl_2_1':
-            return super()._is_embedding_to_invoice_pdf_needed()
-        return False  # ubl must not be embedded to PDF.
```
Comment (smetl) [inline addons/account_edi_ubl/models/account_edi_format.py:266]: I don't understand how such diff is related to l10n_no

File: addons/l10n_nl_edi/models/account_edi_format.py [hunk 6]
```diff
@@ -82,11 +82,9 @@ def _export_nlcius(self, invoice):
     ####################################################
 
     def _check_move_configuration(self, invoice):
-        res = super()._check_move_configuration(invoice)
+        errors = super()._check_move_configuration(invoice)
         if self.code != 'nlcius_1':
-            return res
-
-        errors = []
+            return errors
```
Comment (smetl) [inline addons/l10n_nl_edi/models/account_edi_format.py:87]: this diff is not related to this task, right?

File: addons/l10n_no_edi/data/account_edi_data.xml [hunk 7]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <data>
+        <record id="edi_ehf_3" model="account.edi.format">
+            <field name="name">EHF 3 (Norway)</field>
+            <field name="code">ehf_3</field>
+        </record>
+
+    </data>
```
Comment (smetl) [inline addons/l10n_no_edi/data/account_edi_data.xml:9]: the indent is weird.. one empty line after but not before: make a choice!

File: addons/l10n_no_edi/data/ehf_3_template.xml [hunk 8]
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <data>
+        <template id="export_ehf_3_invoice_partner"
+                  inherit_id="account_edi_ubl_bis3.export_bis3_invoice_partner"
+                  primary="True">
+            <xpath expr="//*[local-name()='PartyTaxScheme']" position="after">
+                <cac:PartyTaxScheme xmlns="urn:oasis:names:specification:ubl:schema:xsd:Invoice-2"
+                xmlns:cac="urn:oasis:names:specification:ubl:schema:xsd:CommonAggregateComponents-2"
+                xmlns:cbc="urn:oasis:names:specification:ubl:schema:xsd:CommonBasicComponents-2">
```
Comment (smetl) [inline addons/l10n_no_edi/data/ehf_3_template.xml:10]: indent is broken

File: addons/l10n_no_edi/data/ehf_3_template.xml [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+        </template>
+
+        <template id="export_ehf_3_invoice"
+                  inherit_id="account_edi_ubl_bis3.export_bis3_invoice"
+                  primary="True">
+            <!-- Only for the supplier, append Foretaksregisteret, see rule NO-R-002 -->
+            <xpath expr="//*[local-name()='AccountingSupplierParty']" position="replace">
+                <cac:AccountingSupplierParty t-call="l10n_no_edi.export_ehf_3_invoice_partner"
+                xmlns="urn:oasis:names:specification:ubl:schema:xsd:Invoice-2"
+                xmlns:cac="urn:oasis:names:specification:ubl:schema:xsd:CommonAggregateComponents-2">
```
Comment (smetl) [inline addons/l10n_no_edi/data/ehf_3_template.xml:26]: indent is broken

File: addons/l10n_no_edi/models/account_edi_format.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+        """
+        is_ubl = super()._is_ubl(filename, tree)
+        return is_ubl and not self._is_ehf_3(filename, tree)
+
+    def _is_ehf_3(self, filename, tree):
+        profile_id = tree.find('./{*}ProfileID')
+        nor_tax = tree.xpath("//*[text()='Foretaksregisteret']")
+        return tree.tag == '{urn:oasis:names:specification:ubl:schema:xsd:Invoice-2}Invoice' and \
+            profile_id is not None and 'peppol' in profile_id.text and \
+            nor_tax is not None
```
Comment (smetl) [inline addons/l10n_no_edi/models/account_edi_format.py:27]: indent is weird!
```
    def _is_ehf_3(self, filename, tree):
        profile_id = tree.find('./{*}ProfileID')
        nor_tax = tree.xpath("//*[text()='Foretaksregisteret']")
        return tree.tag == '{urn:oasis:names:specification:ubl:schema:xsd:Invoice-2}Invoice' \
               and profile_id is not None \
               and 'peppol' in profile_id.text \
               and nor_tax is not None
```
is more readable to me

File: addons/l10n_no_edi/models/account_edi_format.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+
+    ####################################################
+    # Import
+    ####################################################
+
+    def _is_ubl(self, filename, tree):
+        """ OVERRIDE so that the generic ubl parser does not parse BIS3 any longer.
+        """
+        is_ubl = super()._is_ubl(filename, tree)
+        return is_ubl and not self._is_ehf_3(filename, tree)
```
Comment (smetl) [inline addons/l10n_no_edi/models/account_edi_format.py:20]: this is strange... why not
```
if self.code == '<truc_no_edi>':
    return self._is_ehf_3(filename, tree)
else:
    return super()._is_ubl(filename, tree)
```

File: addons/l10n_no_edi/tests/test_ehf.py [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+        '''
+
+    def test_ehf_import(self):
+        invoice = self.env['account.move'].with_context(default_move_type='in_invoice').create({})
+        invoice_count = len(self.env['account.move'].search([]))
+        self.update_invoice_from_file('l10n_no_edi', 'test_xml_file', 'ehf_test.xml', invoice)
+        self.assertEqual(len(self.env['account.move'].search([])), invoice_count)
+        self.assertEqual(invoice.amount_total, 1801.78)
+        self.assertEqual(invoice.amount_tax, 365.28)
+        self.assertEqual(invoice.partner_id, self.partner_a)
```
Comment (smetl) [inline addons/l10n_no_edi/tests/test_ehf.py:215]: use `search_count` & `assertRecordValues`:
```
    def test_ehf_import(self):
        invoice = self.env['account.move'].with_context(default_move_type='in_invoice').create({})
        invoice_count = self.env['account.move'].search_count([])
        self.update_invoice_from_file('l10n_no_edi', 'test_xml_file', 'ehf_test.xml', invoice)
        self.assertEqual(self.env['account.move'].search_count([]), invoice_count)
        self.assertRecordValues(invoice, [{
            'partner_id': self.partner_a.id,
            'amount_total': 1801.78,
            'amount_tax': 365.28,
        }])
```

**Review by smetl — CHANGES_REQUESTED** [review]

File: addons/account_edi_ubl/models/account_edi_format.py (hunk 4 above)
Comment (bfr-o) [reply addons/account_edi_ubl/models/account_edi_format.py:198]: Yes, I rebased

File: addons/account_edi_ubl/models/account_edi_format.py (hunk 5 above)
Comment (bfr-o) [reply addons/account_edi_ubl/models/account_edi_format.py:266]: It's not really. What I'm trying here is to uniformise this kind of code for all ubl formats. Since it does the same thing, it's easier to read if the code is the same.

I did this in a different commit though.

File: addons/l10n_nl_edi/models/account_edi_format.py (hunk 6 above)
Comment (bfr-o) [reply addons/l10n_nl_edi/models/account_edi_format.py:87]: See https://github.com/odoo/odoo/pull/70679#discussion_r680974735

File: addons/l10n_no_edi/models/account_edi_format.py (hunk 11 above)
Comment (bfr-o) [reply addons/l10n_no_edi/models/account_edi_format.py:20]: `self.code` has nothing to do here, it's already checked in `_create_invoice_from_xml_tree` and `_update_invoice_from_xml_tree` that calls this function. 

ubl is more general than ehf, therefore if the parser of ubl goes first, it will parse the file even though it should be parsed by the ehf decoder, to yield a better result. This restricts our detection of generic ubl so it ignores profiles that should be decoded by a dedicated decoder.

**Comment by smetl** [comment]
@robodoo r+ rebase-ff

**Comment by smetl** [comment]
@odoo/rd-security Need a review here.
Markup introduced by https://github.com/odoo/odoo/commit/7df343dd1be1f286493660ed8fd28669ab4a8311

File: addons/l10n_no_edi/models/account_edi_format.py [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+                    'bis3_endpoint': partner.l10n_no_bronnoysund_number,
+                    'bis3_endpoint_scheme': '0192',
+                })
+
+        return values
+
+    def _export_ehf_3(self, invoice):
+        self.ensure_one()
+        # Create file content.
+        xml_content = markupsafe.Markup("<?xml version='1.0' encoding='UTF-8'?>")
```
Comment (xmo-odoo) [inline addons/l10n_no_edi/models/account_edi_format.py:59]: Is the prologue actually needed?

File: addons/l10n_no_edi/models/account_edi_format.py [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+        xml_content += self.env.ref('l10n_no_edi.export_ehf_3_invoice')._render(self._get_ehf_3_values(invoice))
+        vat = invoice.company_id.partner_id.commercial_partner_id.vat
+        xml_name = 'ehf-%s%s%s.xml' % (vat or '', '-' if vat else '', invoice.name.replace('/', '_'))
+        return self.env['ir.attachment'].create({
+            'name': xml_name,
+            'datas': base64.encodebytes(xml_content),
```
Comment (xmo-odoo) [inline addons/l10n_no_edi/models/account_edi_format.py:65]: Or set the `raw` attribute and don't need to unnecessarily b64encode/b64decode. That probably requires explicitely `encode`-ing `xml_content`.

File: addons/l10n_no_edi/models/account_edi_format.py [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+    ####################################################
+
+    def _check_move_configuration(self, invoice):
+        errors = super()._check_move_configuration(invoice)
+        if self.code != 'ehf_3':
+            return errors
+
+        supplier = invoice.company_id.partner_id.commercial_partner_id
+        if supplier.country_code == 'NO' and not supplier.l10n_no_bronnoysund_number:
+            errors.append(_("The supplier %s must have a Bronnoysund company registry.", supplier.display_name))
```
Comment (xmo-odoo) [inline addons/l10n_no_edi/models/account_edi_format.py:82]: `%r` can be useful in this sort of situation as it will automatically quote the string, which can make for a clearer result e.g.

> The supplier the belgian chocolate group must have a Bronnoysund company registry.

versus

> The supplier 'the belgian chocolate group' must have a Bronnoysund company registry.

File: addons/l10n_no_edi/models/account_edi_format.py [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+
+    def _get_ehf_3_values(self, invoice):
+        values = super()._get_bis3_values(invoice)
+        for partner_vals in (values['customer_vals'], values['supplier_vals']):
+            partner = partner_vals['partner']
+            if partner.country_code == 'NO':
+                partner_vals.update({
+                    'bis3_endpoint': partner.l10n_no_bronnoysund_number,
+                    'bis3_endpoint_scheme': '0192',
+                })
```
Comment (xmo-odoo) [inline addons/l10n_no_edi/models/account_edi_format.py:52]: Why use `dict.update` with a literal dict?

File: addons/l10n_no_edi/models/account_edi_format.py [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+
+    def _is_ubl(self, filename, tree):
+        """ OVERRIDE so that the generic ubl parser does not parse BIS3 any longer.
+        """
+        is_ubl = super()._is_ubl(filename, tree)
+        return is_ubl and not self._is_ehf_3(filename, tree)
+
+    def _is_ehf_3(self, filename, tree):
+        profile_id = tree.find('./{*}ProfileID')
+        nor_tax = tree.xpath("//*[text()='Foretaksregisteret']")
```
Comment (xmo-odoo) [inline addons/l10n_no_edi/models/account_edi_format.py:25]: Is this really the only way to detect it?

Incidentally it seems like a waste to create those locals as it's going to perform those sub-searches even if the root is not an UBL invoice.

Also for the profile `findtext` could be better than `find` then dereferencing the `text`.

File: addons/l10n_no_edi/models/account_edi_format.py [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+        return tree.tag == '{urn:oasis:names:specification:ubl:schema:xsd:Invoice-2}Invoice' \
+            and profile_id is not None \
+            and 'peppol' in profile_id.text \
+            and nor_tax is not None
+
+    def _bis3_get_extra_partner_domains(self, tree):
+        if self.code == 'ehf_3':
+            endpoint = tree.find('./{*}AccountingSupplierParty/{*}Party/{*}EndpointID')
```
Comment (xmo-odoo) [inline addons/l10n_no_edi/models/account_edi_format.py:33]: Defining a namespace mapping and using that with prefixes would probably be cleaner than just globbing everything.

File: addons/l10n_no_edi/models/account_edi_format.py [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+            if endpoint is not None:
+                scheme = endpoint.attrib['schemeID']
+                if scheme == '0192' and endpoint.text:
```
Comment (xmo-odoo) [inline addons/l10n_no_edi/models/account_edi_format.py:36]: FWIW using xpath the entire thing should be done at once e.g.
```python
tree.xpath('./cac:AccountingCustomerParty/cac:Party/cbc:EndpointId[@schemeId="0192"]/text()', namespaces=peppol_ns)
```
or something along those lines.

File: addons/l10n_no/models/res_partner.py [hunk 20]
```diff
@@ -0,0 +1,8 @@
+# coding: utf-8
+from odoo import fields, models
+
+
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+
+    l10n_no_bronnoysund_number = fields.Char(string='Bronnoysund company registry', size=9)
```
Comment (xmo-odoo) [inline addons/l10n_no/models/res_partner.py:8]: It's missing the "ø"s. Also the "company registry" seems to be specifically called "Register of Business Enterprises", might be a good idea to use that as the Brønnøysund Register Centre seems to have a *lot* of registers which I assume are mostly not suitable?

Might also be a good idea to use the full name (Brønnøysundregistrene), as Brønnøysund is technically the name of the town where the center is located.

**Review by xmo-odoo — COMMENTED** [review]
A few other comments going through the PR.

File: addons/l10n_no_edi/models/account_edi_format.py (hunk 17 above)
Comment (bfr-o) [reply addons/l10n_no_edi/models/account_edi_format.py:25]: Foretaksregisteret is basically the only difference for the Norwegian profile of BIS3, so yes, it's the only way.
I added the namespaces and better refined the path for nor_tax.
I use `findtext` for ProfileID.
The use of `findtext` is difficult for nor_tax has the same node is present twice and we need the one with the correct text.

File: addons/l10n_no/models/res_partner.py (hunk 20 above)
Comment (bfr-o) [reply addons/l10n_no/models/res_partner.py:8]: I now mention the register of legal entities, and the full name (translated to English) to avoid confusion.

The reference I used is : https://docs.peppol.eu/poacc/billing/3.0/codelist/eas/ (see 0192).

**Review by smetl — APPROVED** [review]
@robodoo r+
